tiledb/cloud/notebook.py

```diff
@@ -35,3 +50,290 @@ def rename_notebook(
+#    different path. No clobbering
+# 2. Auto-increment the array name, e.g. from 'testing-upload' to 'testing-upload-1'
+#    and then 'testing-upload-2' the next time, and so on.
+# 3. Overwrite
+#
+# Thoughts:
+# * Option 3 isn't a safe default -- for those who want it it's fine but for
+#   those who don't it can be seen as unwelcome data loss.
+# * Option 2 is a not-bad default -- there is no data loss, but some users
+#   might be left feeling 'Why are you creating all these versions? I just
+#   want to update one notebook, not have twenty copies."
+# * Option 1 is a safe default -- there is no data loss and no profusion of
+#   copies. However, it is more frictional for the user, requiring them to
+#   make the decision.
+#
+# Implementation:
+#
+# * We could have a force-overwrite argument, optional, default False.
+# * We could have a behavior-on-exist argument, of enum type, 3 cases, one
+#   for each of the options above.
+#
+# Status: As of this writing: we have implemented option 1, and we don't have
+# an overwrite/update-in-place flag.
+def upload_notebook_from_file(
+    ipynb_file_name: str,
+    namespace: str,
+    array_name: str,
+    storage_path: str,
+    storage_credential_name: str,
+) -> str:
+    """
+    Uploads a local-disk notebook file to TileDB Cloud.
+    :param ipnyb_file_name: such as "./mycopy.ipynb". Must be local; no S3 URI
+      support at present.
+    :param namespace: such as "janedoe".
+    :param array_name : name to be seen in the UI, such as "testing-upload".
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param storage_credential_name: such as "janedoe-creds", typically from the
+      user's account settings.
+    :return: TileDB array name, such as "tiledb://janedoe/testing-upload".
+    """
+
+    with open(ipynb_file_name, "r") as handle:
+        ipynb_file_contents = handle.read()
+
+    return upload_notebook_contents(
+        ipynb_file_contents,
+        storage_path,
+        array_name,
+        namespace,
+        storage_credential_name,
+    )
+
+
+def upload_notebook_contents(
+    ipynb_file_contents: str,
+    storage_path: str,
+    array_name: str,
+    namespace: str,
+    storage_credential_name: str,
+) -> str:
+    """
+    Uploads a notebook file to TileDB Cloud.
+    :param ipnyb_file_contents: The contents of the notebook file as a string,
+      nominally in JSON format.
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param array_name : name to be seen in the UI, such as "testing-upload"
+    :param namespace: such as "janedoe".
+    :param storage_credential_name: such as "janedoe-creds", typically from the
+      user's account settings.
+    :return: TileDB array name, such as "tiledb://janedoe/testing-upload".
+    """
+
+    ctx = tiledb.cloud.Ctx(
+        {
```

Comment:
weird that black gave you this. will it let you do

```python
ctx = tiledb.cloud.Ctx(
    {"rest.creation_access_credentials_name": storage_credential_name},
)
```
?

Reply:
yup, black is good with either